grammarinator/resources/codegen/GeneratorTemplate.py.jinja

```diff
@@ -48,9 +48,17 @@ if self._max_depth >= {{ 0 if node.min == 1 else node.min_depth }}:
 
 {% macro processAlternationNode(node) %}
 choice = self._model.choice(current, {{ node.idx }}, [0 if [{{ node.min_depth | join(', ') }}][i] > self._max_depth else w for i, w in enumerate([{{ node.conditions | join(', ') }}])])
-{% set literal_alts = node.literal_alternatives() %}
-{% if literal_alts %}
-UnlexerRule(src=[{{ literal_alts | join(', ')}}][choice], parent=current)
+{% set simple_lits, simple_rules, mixed_alts = node.simple_alternatives() %}
+{% if simple_lits %}
+UnlexerRule(src=[{% for lit in simple_lits %}'{{ lit }}'{% if not loop.last %}, {% endif %}{% endfor %}][choice], parent=current)
+{% elif simple_rules %}
+[{% for rule in simple_rules %}self.{{ rule }}{% if not loop.last %}, {% endif %}{% endfor %}][choice](parent=current)
+{% elif mixed_alts %}
+alt = [{% for alt in mixed_alts %}{% if alt[1] is none %}('{{ alt[0] }}', {{ alt[1] }}){% else %}({{ alt[0] }}, self.{{ alt[1] }}){% endif %}{% if not loop.last %}, {% endif %}{% endfor %}][choice]
+if alt[0] is not None:
+    UnlexerRule(src=alt[0], parent=current)
+else:
+    alt[1](parent=current)
```

Comment:
This template could be adapted with just a bit of reordering:

```py
{% set simple_lits, simple_rules = node.simple_alternatives() %}
{% if simple_lits and simple_rules %}
src = [{% for lit in simple_lits %}{% if lit is not none %}'{{ lit }}'{% else %}None{% endif %}{% if not loop.last %}, {% endif %}{% endfor %}][choice]
rule = [{% for rule in simple_rules %}{% if rule is not none %}self.{{ rule }}{% else %}None{% endif %}{% if not loop.last %}, {% endif %}{% endfor %}][choice]
if src is not None:
    UnlexerRule(src=src, parent=current)
else:
    rule(parent=current)
{% if simple_lits %}
UnlexerRule(src=[{% for lit in simple_lits %}'{{ lit }}'{% if not loop.last %}, {% endif %}{% endfor %}][choice], parent=current)
{% elif simple_rules %}
[{% for rule in simple_rules %}self.{{ rule }}{% if not loop.last %}, {% endif %}{% endfor %}][choice](parent=current)
{% else %}
```

Let me know whether this approach is worth considering.

Reply:
Thanks, it's a cleaner solution indeed.